File: src/machine/units/quota_probe.py
from __future__ import annotations

import json
import os
import re
import select
import shlex
import shutil
import subprocess
import sys
import tempfile
import time
import uuid
from datetime import datetime, timezone
from pathlib import Path

CODEX_APP_SERVER_TIMEOUT_SECONDS = 6.0
CODEX_APP_SERVER_COMMAND = [
    "codex",
    "app-server",
    "--listen",
    "stdio://",
    "--session-source",
    "statusline",
]
CODEX_CLIENT_INFO = {
    "name": "empty-status-probe",
    "version": "0.0.0",
}
CLAUDE_REFRESH_SUCCESS_SECONDS = 15.0 * 60.0
CLAUDE_BOOT_TIMEOUT_SECONDS = 8.0
CLAUDE_VIEW_TIMEOUT_SECONDS = 8.0
CLAUDE_PANE_HEIGHT = 40
CLAUDE_PANE_WIDTH = 120
CLAUDE_STATUS_COMMAND = [
    "claude",
    "--model",
    "haiku",
    "--effort",
    "low",
    "--tools",
    "",
]
CLAUDE_SESSION_USAGE_RE = re.compile(
    r"Current session\s*\n.*?(\d+)% used\s*\n\s*Resets ([^\n]+)",
    re.DOTALL,
)
CLAUDE_WEEKLY_USAGE_RE = re.compile(
    r"Current week \(all models\)\s*\n.*?(\d+)% used\s*\n\s*Resets ([^\n]+)",
    re.DOTALL,
)


def utc_now() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def xdg_cache_home() -> Path:
    raw = os.environ.get("XDG_CACHE_HOME")
    if raw:
        return Path(raw)
    return Path.home() / ".cache"


def claude_cache_path() -> Path:
    return xdg_cache_home() / "empty-status" / "claude-rate-limits.json"


def clamp_percent(value: object) -> int | None:
    if not isinstance(value, (int, float)):
        return None
    return max(0, min(100, int(round(value))))


def parse_rfc3339_age_seconds(raw: str) -> float | None:
    try:
        parsed = datetime.fromisoformat(raw.replace("Z", "+00:00"))
    except ValueError:
        return None
    return max(0.0, (datetime.now(timezone.utc) - parsed).total_seconds())


def newest_project_files(path: Path) -> list[Path]:
    ranked: list[tuple[float, Path]] = []
    try:
        projects = [child for child in path.iterdir() if child.is_dir()]
    except FileNotFoundError:
        return []

    for project in projects:
        try:
            files = [child for child in project.iterdir() if child.is_file() and child.suffix == ".jsonl"]
        except OSError:
            continue
        for file in files:
            try:
                ranked.append((file.stat().st_mtime, file))
            except OSError:
                continue

    ranked.sort(key=lambda item: item[0], reverse=True)
    return [file for _, file in ranked]


def reverse_text_lines(path: Path, block_size: int = 1 << 16):
    with path.open("rb") as handle:
        handle.seek(0, os.SEEK_END)
        offset = handle.tell()
        remainder = b""
        while offset > 0:
            span = min(block_size, offset)
            offset -= span
            handle.seek(offset)
            chunk = handle.read(span)
            lines = (chunk + remainder).split(b"\n")
            remainder = lines[0]
            for raw in reversed(lines[1:]):
                yield raw.decode("utf-8", "replace")
        if remainder:
            yield remainder.decode("utf-8", "replace")


def read_json_rpc_line(handle, timeout_seconds: float) -> dict[str, object] | None:
    ready, _, _ = select.select([handle], [], [], timeout_seconds)
    if not ready:
        return None

    line = handle.readline()
    if not line:
        return None

    try:
        message = json.loads(line)
    except json.JSONDecodeError:
        return None
    return message if isinstance(message, dict) else None


def rpc_call(
    proc: subprocess.Popen[str],
    request_id: int,
    method: str,
    params: dict[str, object],
    timeout_seconds: float,
) -> dict[str, object] | None:
    if proc.stdin is None or proc.stdout is None:
        return None

    proc.stdin.write(
        json.dumps(
            {
                "jsonrpc": "2.0",
                "id": request_id,
                "method": method,
                "params": params,
            },
            separators=(",", ":"),
        )
        + "\n"
    )
    proc.stdin.flush()

    deadline = time.monotonic() + timeout_seconds
    while time.monotonic() < deadline:
        remaining = max(0.0, deadline - time.monotonic())
        message = read_json_rpc_line(proc.stdout, remaining)
        if message is None:
            return None
        if message.get("id") != request_id:
            continue
        result = message.get("result")
        return result if isinstance(result, dict) else None
    return None


def snapshot_field(snapshot: dict[str, object], *names: str) -> object:
    for name in names:
        value = snapshot.get(name)
        if value is not None:
            return value
    return None


def select_window(snapshot: dict[str, object], duration_minutes: int) -> dict[str, object] | None:
    for key in ("primary", "secondary"):
        window = snapshot.get(key)
        if (
            isinstance(window, dict)
            and snapshot_field(window, "window_minutes", "windowDurationMins") == duration_minutes
        ):
            return window
    return None


def read_codex_quota() -> dict[str, object] | None:
    if shutil.which("codex") is None:
        return None

    try:
        proc = subprocess.Popen(
            CODEX_APP_SERVER_COMMAND,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            text=True,
        )
    except OSError:
        return None

    try:
        initialized = rpc_call(
            proc,
            1,
            "initialize",
            {
                "clientInfo": CODEX_CLIENT_INFO,
                "capabilities": {"experimentalApi": True},
            },
            CODEX_APP_SERVER_TIMEOUT_SECONDS,
        )
        if initialized is None:
            return None

        result = rpc_call(
            proc,
            2,
            "account/rateLimits/read",
            {},
            CODEX_APP_SERVER_TIMEOUT_SECONDS,
        )
        if result is None:
            return None

        snapshots = snapshot_field(result, "rateLimitsByLimitId")
        if isinstance(snapshots, dict):
            snapshot = snapshots.get("codex")
            if not isinstance(snapshot, dict):
                return None
        else:
            snapshot = snapshot_field(result, "rateLimits")
            if not isinstance(snapshot, dict):
                return None

        weekly = select_window(snapshot, 10080)
        if weekly is None:
            return None
        five_hour = select_window(snapshot, 300)
        weekly_used = clamp_percent(snapshot_field(weekly, "used_percent", "usedPercent"))
        if weekly_used is None:
            return None
        return {
            "captured_at": utc_now(),
            "weekly_used_percent": weekly_used,
            "weekly_resets_at": snapshot_field(weekly, "resets_at", "resetsAt"),
            "five_hour_used_percent": clamp_percent(
                None
                if five_hour is None
                else snapshot_field(five_hour, "used_percent", "usedPercent")
            ),
            "five_hour_resets_at": None
            if five_hour is None
            else snapshot_field(five_hour, "resets_at", "resetsAt"),
            "plan_type": snapshot_field(snapshot, "plan_type", "planType"),
        }
    except (BrokenPipeError, OSError, subprocess.SubprocessError):
        return None
    finally:
        if proc.stdin is not None:
            proc.stdin.close()
        if proc.poll() is None:
            proc.terminate()
            try:
                proc.wait(timeout=1.0)
            except subprocess.TimeoutExpired:
                proc.kill()
                try:
                    proc.wait(timeout=1.0)
                except subprocess.TimeoutExpired:
                    pass


def load_claude_cache() -> dict[str, object] | None:
    target = claude_cache_path()
    try:
        parsed = json.loads(target.read_text(encoding="utf-8"))
    except (FileNotFoundError, OSError, json.JSONDecodeError):
        return None
    return parsed if isinstance(parsed, dict) else None


def write_claude_cache(snapshot: dict[str, object]) -> None:
    target = claude_cache_path()
    target.parent.mkdir(parents=True, exist_ok=True)
    temp = target.with_suffix(".json.tmp")
    temp.write_text(json.dumps(snapshot, separators=(",", ":")), encoding="utf-8")
    temp.replace(target)


def read_recent_claude_cwd() -> Path | None:
    root = Path.home() / ".claude" / "projects"
    probed_files = 0
    for file in newest_project_files(root):
        probed_files += 1
        if probed_files > 96:
            return None
        try:
            for line in reverse_text_lines(file):
                if '"cwd"' not in line:
                    continue
                event = json.loads(line)
                cwd = event.get("cwd")
                if not isinstance(cwd, str):
                    continue
                path = Path(cwd)
                if path.is_dir():
                    return path
        except (OSError, UnicodeDecodeError, json.JSONDecodeError):
            continue
    return None


def run_tmux(base: list[str], cwd: Path, *args: str, capture: bool = False) -> str:
    result = subprocess.run(
        [*base, *args],
        check=True,
        cwd=cwd,
        env={**os.environ, "TMUX_TMPDIR": tempfile.gettempdir()},
        stdout=subprocess.PIPE if capture else subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        text=True,
    )
    return "" if result.stdout is None else result.stdout


def wait_for_tmux_pane(
    base: list[str],
    cwd: Path,
    predicate,
    timeout_seconds: float,
) -> str:
    import time

    deadline = time.monotonic() + timeout_seconds
    last = ""
    while time.monotonic() < deadline:
        last = run_tmux(base, cwd, "capture-pane", "-p", "-S", "-", capture=True)
        if predicate(last):
            return last
        time.sleep(0.25)
    return last


def parse_claude_usage_pane(pane: str) -> dict[str, object] | None:
    session_match = CLAUDE_SESSION_USAGE_RE.search(pane)
    weekly_match = CLAUDE_WEEKLY_USAGE_RE.search(pane)
    if session_match is None or weekly_match is None:
        return None

    five_hour_used = clamp_percent(int(session_match.group(1)))
    weekly_used = clamp_percent(int(weekly_match.group(1)))
    if five_hour_used is None or weekly_used is None:
        return None

    return {
        "captured_at": utc_now(),
        "five_hour_used_percent": five_hour_used,
        "five_hour_resets_at": session_match.group(2).strip(),
        "weekly_used_percent": weekly_used,
        "weekly_resets_at": weekly_match.group(2).strip(),
    }


def probe_claude_quota() -> dict[str, object] | None:
    if shutil.which("claude") is None or shutil.which("tmux") is None:
        return None

    cwd = read_recent_claude_cwd()
    if cwd is None:
        return None

    socket_dir = Path(tempfile.mkdtemp(prefix="empty-status-quota-"))
    socket_path = socket_dir / f"{uuid.uuid4().hex}.sock"
    base = ["tmux", "-S", str(socket_path)]

    try:
        run_tmux(
            base,
            cwd,
            "new-session",
            "-d",
            "-x",
            str(CLAUDE_PANE_WIDTH),
            "-y",
            str(CLAUDE_PANE_HEIGHT),
            shlex.join(CLAUDE_STATUS_COMMAND),
        )
        if "❯" not in wait_for_tmux_pane(
            base,
            cwd,
            lambda pane: "❯" in pane,
            CLAUDE_BOOT_TIMEOUT_SECONDS,
        ):
            return None

        run_tmux(base, cwd, "send-keys", "/status", "Enter")
        wait_for_tmux_pane(
            base,
            cwd,
            lambda pane: "Version:" in pane and "Status" in pane,
            CLAUDE_VIEW_TIMEOUT_SECONDS,
        )

        run_tmux(base, cwd, "send-keys", "Right")
        wait_for_tmux_pane(
            base,
            cwd,
            lambda pane: "Search settings..." in pane,
            CLAUDE_VIEW_TIMEOUT_SECONDS,
        )

        run_tmux(base, cwd, "send-keys", "Right")
        pane = wait_for_tmux_pane(
            base,
            cwd,
            lambda pane: "Current week (all models)" in pane,
            CLAUDE_VIEW_TIMEOUT_SECONDS,
        )
        snapshot = parse_claude_usage_pane(pane)
        if snapshot is not None:
            write_claude_cache(snapshot)
            return snapshot
        return load_claude_cache()
    except (OSError, subprocess.SubprocessError):
        return load_claude_cache()
    finally:
        subprocess.run(
            [*base, "kill-server"],
            cwd=cwd,
            env={**os.environ, "TMUX_TMPDIR": tempfile.gettempdir()},
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            check=False,
            text=True,
        )
        shutil.rmtree(socket_dir, ignore_errors=True)


def read_claude_quota(force_refresh: bool) -> dict[str, object] | None:
    cached = load_claude_cache()
    if not force_refresh and cached is not None:
        captured_at = cached.get("captured_at")
        if isinstance(captured_at, str):
            age_seconds = parse_rfc3339_age_seconds(captured_at)
            if age_seconds is not None and age_seconds < CLAUDE_REFRESH_SUCCESS_SECONDS:
                return cached

    refreshed = probe_claude_quota()
    if refreshed is not None:
        return refreshed
    return cached


def main() -> None:
    force_refresh = "--force" in sys.argv[1:]
    snapshot = {
        "sampled_at": utc_now(),
        "codex": read_codex_quota(),
        "claude": read_claude_quota(force_refresh),
    }
    print(json.dumps(snapshot, separators=(",", ":")), flush=True)


if __name__ == "__main__":
    main()

File: src/machine/units/quota_probe.rs
use crate::units::quota::{ClaudeQuota, CodexQuota, ProbeSnapshot, QuotaProvider, QuotaProviders};
use chrono::{SecondsFormat, Utc};
use serde::{Deserialize, Serialize, de::DeserializeOwned};
use std::collections::HashMap;
use std::ffi::OsStr;
use std::fs;
use std::io::{Read, Write};
use std::path::PathBuf;
use std::process::Stdio;
use std::time::{Duration, Instant};
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader, Lines};
use tokio::process::{ChildStdin, ChildStdout, Command};
use tokio::time::timeout;

pub(crate) const PROBE_ARG: &str = "--empty-status-quota-probe";

pub(crate) const PROVIDER_ARG: &str = "--provider";
const CODEX_APP_SERVER_TIMEOUT: Duration = Duration::from_secs(6);
const CODEX_APP_SERVER_ARGS: &[&str] = &["app-server", "--listen", "stdio://"];
const CODEX_CLIENT_NAME: &str = "empty-status-probe";

#[derive(Debug, Clone)]
pub(crate) struct ProbeArgs {
    providers: QuotaProviders,
}

impl ProbeArgs {
    pub(crate) fn parse<I, S>(args: I) -> anyhow::Result<Option<Self>>
    where
        I: IntoIterator<Item = S>,
        S: AsRef<OsStr>,
    {
        let mut probe = false;
        let mut providers = Vec::new();
        let mut args = args.into_iter();
        while let Some(arg) = args.next() {
            let arg = arg.as_ref();
            if arg == OsStr::new(PROBE_ARG) {
                probe = true;
            } else if arg == OsStr::new(PROVIDER_ARG) {
                let Some(provider) = args.next() else {
                    anyhow::bail!("{PROVIDER_ARG} requires a provider");
                };
                let raw = provider.as_ref().to_string_lossy();
                let Some(provider) = QuotaProvider::from_arg(&raw) else {
                    anyhow::bail!("unknown quota provider `{raw}`");
                };
                providers.push(provider);
            }
        }

        if !probe {
            return Ok(None);
        }

        let providers = if providers.is_empty() {
            QuotaProviders::default()
        } else {
            QuotaProviders::new(providers)?
        };
        Ok(Some(Self { providers }))
    }
}

pub(crate) async fn run(args: ProbeArgs) -> anyhow::Result<()> {
    let snapshot = probe_quota(&args.providers).await;
    let line = serde_json::to_string(&snapshot)?;
    writeln!(std::io::stdout().lock(), "{line}")?;
    Ok(())
}

pub(crate) fn run_claude_statusline() -> anyhow::Result<()> {
    let raw = read_stdin_json_object()?;
    let payload = serde_json::from_str::<ClaudeStatuslinePayload>(&raw).unwrap_or_default();
    let quota = payload.quota(utc_now());
    if let Some(quota) = quota.as_ref() {
        let _ = write_claude_cache(quota);
    }
    print!("{}", render_claude_statusline(quota.as_ref(), &payload));
    Ok(())
}

fn read_stdin_json_object() -> anyhow::Result<String> {
    let mut stdin = std::io::stdin().lock();
    let mut raw = Vec::new();
    let mut byte = [0_u8; 1];
    while stdin.read(&mut byte)? == 1 {
        raw.push(byte[0]);
        if json_object_end(&raw) == Some(raw.len()) {
            break;
        }
    }
    Ok(String::from_utf8(raw)?)
}

fn json_object_end(raw: &[u8]) -> Option<usize> {
    let mut started = false;
    let mut string = false;
    let mut escape = false;
    let mut depth = 0_usize;

    for (index, byte) in raw.iter().copied().enumerate() {
        if !started {
            if byte.is_ascii_whitespace() {
                continue;
            }
            if byte != b'{' {
                return None;
            }
            started = true;
            depth = 1;
            continue;
        }

        if string {
            if escape {
                escape = false;
            } else if byte == b'\\' {
                escape = true;
            } else if byte == b'"' {
                string = false;
            }
            continue;
        }

        match byte {
            b'"' => string = true,
            b'{' => depth += 1,
            b'}' => {
                depth = depth.checked_sub(1)?;
                if depth == 0 {
                    return Some(index + 1);
                }
            }
            _ => {}
        }
    }

    None
}

async fn probe_quota(providers: &QuotaProviders) -> ProbeSnapshot {
    let read_codex = async {
        if providers.contains(QuotaProvider::Codex) {
            read_codex_quota().await
        } else {
            None
        }
    };
    let read_claude = async {
        if providers.contains(QuotaProvider::Claude) {
            read_claude_quota().await
        } else {
            None
        }
    };
    let (codex, claude) = tokio::join!(read_codex, read_claude);
    ProbeSnapshot { codex, claude }
}

async fn read_codex_quota() -> Option<CodexQuota> {
    let mut command = Command::new("codex");
    let _ = command
        .args(CODEX_APP_SERVER_ARGS)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .kill_on_drop(true);

    let mut child = command.spawn().ok()?;
    let Some(stdin) = child.stdin.take() else {
        reap_codex_child(child).await;
        return None;
    };
    let Some(stdout) = child.stdout.take() else {
        reap_codex_child(child).await;
        return None;
    };

    let quota = read_codex_quota_rpc(stdin, stdout).await;
    reap_codex_child(child).await;
    quota
}

async fn reap_codex_child(mut child: tokio::process::Child) {
    let _ = child.start_kill();
    let _ = timeout(Duration::from_secs(1), child.wait()).await;
}

async fn read_codex_quota_rpc(stdin: ChildStdin, stdout: ChildStdout) -> Option<CodexQuota> {
    let mut rpc = JsonRpcClient::new(stdin, stdout);
    let _: serde_json::Value = rpc
        .call(
            1,
            "initialize",
            InitializeParams {
                client_info: ClientInfo {
                    name: CODEX_CLIENT_NAME,
                    version: env!("CARGO_PKG_VERSION"),
                },
                capabilities: InitializeCapabilities {
                    experimental_api: true,
                },
            },
        )
        .await?;

    let envelope: CodexRateLimitsEnvelope = rpc
        .call(2, "account/rateLimits/read", EmptyParams {})
        .await?;
    envelope.quota(utc_now())
}

#[derive(Debug)]
struct JsonRpcClient {
    stdin: ChildStdin,
    lines: Lines<BufReader<ChildStdout>>,
}

impl JsonRpcClient {
    fn new(stdin: ChildStdin, stdout: ChildStdout) -> Self {
        Self {
            stdin,
            lines: BufReader::new(stdout).lines(),
        }
    }

    async fn call<P, T>(&mut self, id: u64, method: &str, params: P) -> Option<T>
    where
        P: Serialize,
        T: DeserializeOwned,
    {
        let request = JsonRpcRequest {
            jsonrpc: "2.0",
            id,
            method,
            params,
        };
        let mut encoded = serde_json::to_vec(&request).ok()?;
        encoded.push(b'\n');
        self.stdin.write_all(&encoded).await.ok()?;
        self.stdin.flush().await.ok()?;

        let deadline = Instant::now() + CODEX_APP_SERVER_TIMEOUT;
        loop {
            let remaining = deadline.checked_duration_since(Instant::now())?;
            let line = match timeout(remaining, self.lines.next_line()).await {
                Ok(Ok(Some(line))) => line,
                Ok(Ok(None)) | Ok(Err(_)) | Err(_) => return None,
            };
            let message = serde_json::from_str::<serde_json::Value>(&line).ok()?;
            if message.get("id").and_then(serde_json::Value::as_u64) != Some(id) {
                continue;
            }
            return serde_json::from_value::<JsonRpcResponse<T>>(message)
                .ok()?
                .result;
        }
    }
}

#[derive(Debug, Serialize)]
struct JsonRpcRequest<'a, P> {
    jsonrpc: &'static str,
    id: u64,
    method: &'a str,
    params: P,
}

#[derive(Debug, Deserialize)]
struct JsonRpcResponse<T> {
    result: Option<T>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct InitializeParams<'a> {
    client_info: ClientInfo<'a>,
    capabilities: InitializeCapabilities,
}

#[derive(Debug, Serialize)]
struct ClientInfo<'a> {
    name: &'a str,
    version: &'a str,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct InitializeCapabilities {
    experimental_api: bool,
}

#[derive(Debug, Serialize)]
struct EmptyParams {}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct CodexRateLimitsEnvelope {
    rate_limits: Option<CodexLimitSnapshot>,
    rate_limits_by_limit_id: Option<HashMap<String, CodexLimitSnapshot>>,
}

impl CodexRateLimitsEnvelope {
    fn quota(&self, captured_at: String) -> Option<CodexQuota> {
        let snapshot = self
            .rate_limits_by_limit_id
            .as_ref()
            .and_then(|snapshots| snapshots.get("codex"))
            .or(self.rate_limits.as_ref())?;
        let weekly = snapshot.window(7 * 24 * 60)?;
        Some(CodexQuota {
            captured_at,
            weekly_used_percent: percent(weekly.used_percent?)?,
            weekly_resets_at: weekly.resets_at.and_then(EpochSeconds::into_i64),
        })
    }
}

#[derive(Debug, Deserialize)]
struct CodexLimitSnapshot {
    primary: Option<CodexRateWindow>,
    secondary: Option<CodexRateWindow>,
}

impl CodexLimitSnapshot {
    fn window(&self, minutes: u64) -> Option<&CodexRateWindow> {
        self.primary
            .as_ref()
            .into_iter()
            .chain(self.secondary.as_ref())
            .find(|window| window.window_duration_mins == Some(minutes))
    }
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct CodexRateWindow {
    #[serde(alias = "window_minutes")]
    window_duration_mins: Option<u64>,
    #[serde(alias = "used_percent")]
    used_percent: Option<f64>,
    #[serde(alias = "resets_at")]
    resets_at: Option<EpochSeconds>,
}

#[derive(Debug, Clone, Copy, Deserialize)]
#[serde(untagged)]
enum EpochSeconds {
    Integer(i64),
    Float(f64),
}

impl EpochSeconds {
    fn into_i64(self) -> Option<i64> {
        match self {
            Self::Integer(value) => Some(value),
            Self::Float(value) if value.is_finite() => Some(value.round() as i64),
            Self::Float(_) => None,
        }
    }
}

async fn read_claude_quota() -> Option<ClaudeQuota> {
    load_claude_cache()
}

fn load_claude_cache() -> Option<ClaudeQuota> {
    let raw = fs::read_to_string(claude_cache_path()?).ok()?;
    serde_json::from_str(&raw).ok()
}

fn write_claude_cache(snapshot: &ClaudeQuota) -> Option<()> {
    let target = claude_cache_path()?;
    let parent = target.parent()?;
    fs::create_dir_all(parent).ok()?;
    let temp = target.with_extension("json.tmp");
    fs::write(&temp, serde_json::to_vec(snapshot).ok()?).ok()?;
    fs::rename(temp, target).ok()?;
    Some(())
}

fn claude_cache_path() -> Option<PathBuf> {
    let root = nonempty_env_path("XDG_CACHE_HOME")
        .or_else(|| home_dir().map(|home| home.join(".cache")))?;
    Some(root.join("empty-status").join("claude-rate-limits.json"))
}

fn home_dir() -> Option<PathBuf> {
    nonempty_env_path("HOME")
}

fn nonempty_env_path(name: &str) -> Option<PathBuf> {
    let value = std::env::var_os(name)?;
    (!value.is_empty()).then(|| PathBuf::from(value))
}

fn percent(value: f64) -> Option<u8> {
    value
        .is_finite()
        .then(|| value.round().clamp(0.0, 100.0) as u8)
}

fn utc_now() -> String {
    Utc::now().to_rfc3339_opts(SecondsFormat::Secs, true)
}

#[derive(Debug, Default, Deserialize)]
struct ClaudeStatuslinePayload {
    rate_limits: Option<ClaudeStatuslineRateLimits>,
    context_window: Option<ClaudeContextWindow>,
}

impl ClaudeStatuslinePayload {
    fn quota(&self, captured_at: String) -> Option<ClaudeQuota> {
        let rate_limits = self.rate_limits.as_ref()?;
        let five_hour = rate_limits.five_hour.as_ref()?;
        let weekly = rate_limits.seven_day.as_ref()?;
        Some(ClaudeQuota {
            captured_at,
            weekly_used_percent: percent(weekly.used_percentage?)?,
            weekly_resets_at: weekly.resets_at?.into_i64()?,
            five_hour_used_percent: percent(five_hour.used_percentage?)?,
            five_hour_resets_at: five_hour.resets_at?.into_i64()?,
        })
    }
}

#[derive(Debug, Deserialize)]
struct ClaudeStatuslineRateLimits {
    five_hour: Option<ClaudeStatuslineWindow>,
    seven_day: Option<ClaudeStatuslineWindow>,
}

#[derive(Debug, Deserialize)]
struct ClaudeStatuslineWindow {
    #[serde(alias = "usedPercentage")]
    used_percentage: Option<f64>,
    #[serde(alias = "resetsAt")]
    resets_at: Option<EpochSeconds>,
}

#[derive(Debug, Deserialize)]
struct ClaudeContextWindow {
    #[serde(alias = "remainingPercentage")]
    remaining_percentage: Option<f64>,
}

fn render_claude_statusline(
    quota: Option<&ClaudeQuota>,
    payload: &ClaudeStatuslinePayload,
) -> String {
    if let Some(quota) = quota {
        return format!(
            "cc 7d {:>3}% 5h {:>3}%",
            100_u8.saturating_sub(quota.weekly_used_percent),
            100_u8.saturating_sub(quota.five_hour_used_percent)
        );
    }

    payload
        .context_window
        .as_ref()
        .and_then(|context| percent(context.remaining_percentage?))
        .map_or_else(String::new, |remaining| format!("ctx {remaining:>3}%"))
}

#[cfg(test)]
mod tests {
    use super::{
        ClaudeStatuslinePayload, CodexRateLimitsEnvelope, json_object_end, render_claude_statusline,
    };

    #[test]
    fn codex_rate_limit_parser_accepts_current_app_server_shape() {
        let raw = r#"{
            "rateLimits": {
                "limitId": "codex",
                "planType": "pro",
                "primary": {"resetsAt": 1776733749, "usedPercent": 20, "windowDurationMins": 300},
                "secondary": {"resetsAt": 1776967090, "usedPercent": 64, "windowDurationMins": 10080},
                "credits": null
            },
            "rateLimitsByLimitId": {
                "codex": {
                    "limitId": "codex",
                    "planType": "pro",
                    "primary": {"resetsAt": 1776733749, "usedPercent": 20, "windowDurationMins": 300},
                    "secondary": {"resetsAt": 1776967090, "usedPercent": 64, "windowDurationMins": 10080},
                    "credits": null
                }
            }
        }"#;
        let parsed = serde_json::from_str::<CodexRateLimitsEnvelope>(raw)
            .ok()
            .and_then(|envelope| envelope.quota("2026-04-20T00:00:00Z".to_string()));

        assert_eq!(
            parsed.as_ref().map(|quota| quota.weekly_used_percent),
            Some(64)
        );
        assert_eq!(
            parsed.and_then(|quota| quota.weekly_resets_at),
            Some(1_776_967_090)
        );
    }

    #[test]
    fn claude_statusline_payload_reads_session_and_week_windows() {
        let raw = r#"{
            "rate_limits": {
                "five_hour": {"used_percentage": 12, "resets_at": 1776733749},
                "seven_day": {"used_percentage": 37, "resets_at": 1776967090}
            },
            "context_window": {"remaining_percentage": 88}
        }"#;
        let parsed = serde_json::from_str::<ClaudeStatuslinePayload>(raw)
            .ok()
            .and_then(|payload| payload.quota("2026-04-20T00:00:00Z".to_string()));

        assert_eq!(
            parsed.as_ref().map(|quota| quota.five_hour_used_percent),
            Some(12)
        );
        assert_eq!(
            parsed.as_ref().map(|quota| quota.weekly_used_percent),
            Some(37)
        );
        assert_eq!(
            parsed.as_ref().map(|quota| quota.five_hour_resets_at),
            Some(1_776_733_749)
        );
        assert_eq!(
            parsed.as_ref().map(|quota| quota.weekly_resets_at),
            Some(1_776_967_090)
        );
    }

    #[test]
    fn statusline_stdin_reader_stops_after_one_json_object() {
        let raw = br#"  {"outer":{"inner":"not } done","escape":"\\"}} trailing"#;
        assert_eq!(json_object_end(raw), Some(48));
    }

    #[test]
    fn claude_statusline_renders_quota_summary() {
        let raw = r#"{
            "rate_limits": {
                "five_hour": {"used_percentage": 12, "resets_at": 1776733749},
                "seven_day": {"used_percentage": 37, "resets_at": 1776967090}
            }
        }"#;
        let payload = serde_json::from_str::<ClaudeStatuslinePayload>(raw);
        assert!(payload.is_ok());
        let Ok(payload) = payload else {
            return;
        };
        let quota = payload.quota("2026-04-20T00:00:00Z".to_string());
        assert_eq!(
            render_claude_statusline(quota.as_ref(), &payload),
            "cc 7d  63% 5h  88%"
        );
    }
}
